File: Python/Binary Tree.py
#########################
# Binary Tree Structure #
#########################

class BinaryTree:
	def __init__(self, data):
		self.data = data
		self.left = None
		self.right = None

	def display_tree(self, level=0, prefix="Root: "):
		if root is not None:
			if level == 0:
				print(f"{prefix}{str(self.data)}")
			else:
				line = " " * (level * 4) + "|-- "
				if prefix.startswith("L: "):
					line = " " * (level * 4) + "|--L: "
				elif prefix.startswith("R: "):
					line = " " * (level * 4) + "|--R: "
				print(line + str(self.data))

			if self.left is not None:
				self.left.display_tree(level + 1, "L: ")
			if self.right is not None:
				self.right.display_tree(level + 1, "R: ")

	# Recursive
	def inorder_traversal(self):
		# left --> root --> right
		result = []

		if self is None:
			return result

		if self.left:
			result.extend(self.left.inorder_traversal())

		result.append(self.data)

		if self.right:
			result.extend(self.right.inorder_traversal())

		return result

	# Iterative
	def iterative_inorder_traversal(self):
		# left --> root --> right
		if self is None:
			return []

		result = []
		stack = []
		current = self

		while current or stack:
			while current:
				stack.append(current)
				current = current.left

			current = stack.pop()
			result.append(current.data)
			current = current.right

		return result


	# Recursive
	def preorder_traversal(self):
		# root --> left --> right
		if self is None:
			return []

		result = []

		result.append(self.data)

		if self.left:
			result.extend(self.left.preorder_traversal())

		if self.right:
			result.extend(self.right.preorder_traversal())

		return result

	# Iterative
	def iterative_preorder_traversal(self):
		# root --> left --> right
		if self is None:
			return []

		stack = []
		result = []
		current = self

		while current or stack:
			while current:
				result.append(current.data)
				stack.append(current)
				current = current.left

			current = stack.pop()
			current = current.right

		return result

	
	# Recursive
	def postorder_traversal(self):
		# left --> right --> root
		if self is None:
			return []

		result = []

		if self.left:
			result.extend(self.left.postorder_traversal())

		if self.right:
			result.extend(self.right.postorder_traversal())

		result.append(self.data)

		return result

	# Iterative
	def iterative_postorder_traversal(self):
		# left --> right --> root
		if self is None:
			return []

		stack = []
		result = []
		current = self
		prev = None

		while current or stack:
			while current:
				stack.append(current)
				current = current.left

			current = stack[-1]

			if current.right is None or current.right == prev:
				result.append(current.data)
				stack.pop()
				prev = current
				current = None
			else:
				current = current.right

		return result

	def is_same_tree(self, p, q):
		if p is None and q is None:
			return True

		if p is None or q is None:
			return False

		return p.data == q.data and self.is_same_tree(p.left, q.left) and self.is_same_tree(p.right, q.right)


	def is_symmetric(self):
		def is_mirror(left, right):
			if not left and not right:
				return True
			if not left or not right:
				return False
			return (left.data == right.data) and is_mirror(left.left, right.right) and is_mirror(left.right, right.left)

		if not self:
			return True

		return is_mirror(self.left, self.right)


	def is_balanced(self):
		def check_balance(node):
			# An empty tree is balanced and has a height of 0.
			if not node:
				return True, 0

			# Check if the left subtree is balanced and get its height.
			is_left_balanced, left_height = check_balance(node.left)
			if not is_left_balanced:
				return False, 0
				# If the left subtree is not balanced, the whole tree is unbalanced.

			is_right_balanced, right_height = check_balance(node.right)
			if not is_right_balanced:
				return False, 0

			is_current_balanced = abs(left_height - right_height) <= 1
			current_height = max(left_height, right_height) + 1

			return is_current_balanced, current_height

		is_balanced, _ = check_balance(self)
		return is_balanced


	def min_depth(self):
		# Breadth First Search
		if not self:
			return 0

		queue = [(root, 1)]

		while queue:
			node, depth = queue.pop(0)

			if not node.left and not node.right:
				return depth

			if node.left:
				queue.append((node.left, depth + 1))


			if node.right:
				queue.append((node.right, depth + 1))


	def max_depth(self):
		stack = [[self, 1]]
		result = 0

		while stack:
			node, depth = stack.pop()

			if node:
				result = max(result, depth)
				stack.append([node.left, depth + 1])
				stack.append([node.right, depth + 1])
		return result


	def has_path_sum(self, root, target_sum):
		if root is None:
			return False

		target_sum -= root.data

		if root.left is None and root.right is None:
			return target_sum == 0

		return self.has_path_sum(root.left, target_sum) or self.has_path_sum(root.right, target_sum)


	def largest_values(self):
		if not self:
			return []

		result = []
		queue = [self]

		while queue:
			level_max = float('-inf')
			level_size = len(queue)

			for i in range(level_size):
				node = queue.pop(0)
				level_max = max(level_max, node.data)

				if node.left:
					queue.append(node.left)

				if node.right:
					queue.append(node.right)

			result.append(level_max)

		return result


	def binary_tree_paths(self):
		def dfs(node, current_path):
			if not node:
				return

			current_path.append(str(node.data))

			if not node.left and not node.right:
				result.append("-> ".join(current_path))

			dfs(node.left, current_path)
			dfs(node.right, current_path)

			current_path.pop()

		result = []
		dfs(self, [])
		return result

	def count_nodes(self):
	    if not self:
	        return 0

	    depth_left, depth_right = 0, 0
	    left = right = self

	    while left:
	        left = left.left
	        depth_left += 1

	    while right:
	        right = right.right
	        depth_right += 1

	    if depth_left == depth_right:
	        return 2 ** depth_left - 1
	    else:
	        left_count = self.left.count_nodes() if self.left else 0
	        right_count = self.right.count_nodes() if self.right else 0
	        return 1 + left_count + right_count

	def sum_of_left_leaves(self):
	    if not self:
	        return 0

	    left_sum = 0

	    if self.left and not self.left.left and not self.left.right:
	        left_sum += self.left.data

	    left_sum += self.left.sum_of_left_leaves() if self.left else 0
	    left_sum += self.right.sum_of_left_leaves() if self.right else 0

	    return left_sum

	def level_order(self):
		if not self:
			return []

		result = []
		queue = [self]

		while queue:
			level = []
			level_size = len(queue)

			for _ in range(level_size):
				node = queue.pop(0)
				level.append(node.data)

				if node.left:
					queue.append(node.left)

				if node.right:
					queue.append(node.right)

			result.append(level)
		return result
		

#    5
#   / \
#  6   8
#   \   \
#    9   11
#   / \
#  14  15


# Main Root
root = BinaryTree(5)

# 1st Level
root.left = BinaryTree(6)
root.right = BinaryTree(8)

# 2nd Level
root.left.right = BinaryTree(9) 
root.right.right = BinaryTree(11)

# 3rd Level
root.left.right.left = BinaryTree(14)
root.left.right.right = BinaryTree(15)


print("**Binary Tree Sturcture**", "\n")
# InOrder Traversal
print("InOrder Recursive:", root.inorder_traversal())
print("InOrder Iterative:", root.iterative_inorder_traversal())
print("\r")
# PreOrder Traversal
print("PreOrder Recursive:", root.preorder_traversal())
print("PreOrder Iterative:", root.iterative_preorder_traversal())
print("\r")
# PostOrder Traversal
print("PostOrder Recursive:", root.postorder_traversal())
print("PostOrder Iterative:", root.iterative_postorder_traversal())
print("\n")


# Comparisons
# P - Tree
p_root = BinaryTree(1)
p_root.left = BinaryTree(2)
p_root.right = BinaryTree(3)

# Q - Tree
q_root = BinaryTree(1)
q_root.left = BinaryTree(2)
q_root.right = BinaryTree(3)

p_root.display_tree()
q_root.display_tree()

print(p_root.is_same_tree(p_root, q_root))


# Is Symmetric?
#          1
# 		 /   \
# 		2     2
#      / \   / \
#     3   4 4   3


# Main Root
symm_root = BinaryTree(1)

# 1st Level
symm_root.left = BinaryTree(2)
symm_root.right = BinaryTree(2)

# 2nd Level
symm_root.left.left = BinaryTree(3)
symm_root.right.right = BinaryTree(3)
symm_root.left.right = BinaryTree(4)
symm_root.right.left = BinaryTree(4)

print("Symmetric:", symm_root.is_symmetric())

# Is Balanced?
#  - a tree is balance if left & right has <= 1 height difference.
print("Balanced:", symm_root.is_balanced())

# Minimum Depth till a leaf node is met.
print("Min Depth:", symm_root.min_depth())

# Maximum Depth till a leaf node is met.
print("Max Depth:", symm_root.max_depth())


# Path Sum
#        5
#       / \
#      4   8
#     /   / \
#    11  13  4
#   /  \      \
#  7    2      1

# 5 + 4 + 11 + 2 = 22
# target_sum = 22

# Main Root
root_sum = BinaryTree(5)

# 1st Level
root_sum.left = BinaryTree(4)
root_sum.right = BinaryTree(8)

# 2nd Level
root_sum.left.left = BinaryTree(11) 
root_sum.right.left = BinaryTree(13) 
root_sum.right.right = BinaryTree(4) 

# 3rd Level
root_sum.left.left.left = BinaryTree(7)
root_sum.left.left.right = BinaryTree(2)
root_sum.right.right.right = BinaryTree(1)

# Will return True
print(root_sum.has_path_sum(root_sum, 22))

print(root_sum.largest_values())

print(root_sum.binary_tree_paths())



#  Complete Tree
#         1
#        / \
#		2 	3
# 	   /\  /
# 	  4  5 6


# Main Root
complete_root = BinaryTree(1)

# 1st Level
complete_root.left = BinaryTree(2)
complete_root.right = BinaryTree(3)

# 2nd Level
complete_root.left.left = BinaryTree(4)
complete_root.left.right = BinaryTree(5)
complete_root.right.left = BinaryTree(6)

# Counts complete tree nodes
print("Nodes: ", complete_root.count_nodes())

# Sum of all left leaf nodes
#  4 + 6 = 10
print("Left leaf sum: ", complete_root.sum_of_left_leaves())


# Level Order Traversal
print("Level Order Traversal:", complete_root.level_order())

File: C#/Binary Tree.cs
using System;
using System.Linq;
using System.Collections.Generic;

namespace Binary_Tree_Practice
{
    class Program
    {
        static void Main(string[] args)
        {
            //  BINARY TREE STRUCTURE
            /*
                    5
                   / \
                  6   8
                   \   \
                    9   11
                   / \
                  14  15 
            */

            //Main Root
            TreeNode root = new TreeNode(5);

            //1st Level
            root.Left = new TreeNode(6);
            root.Right = new TreeNode(8);

            //2nd Level
            root.Left.Right = new TreeNode(9);
            root.Right.Right = new TreeNode(11);

            //3rd Level
            root.Left.Right.Left = new TreeNode(14);
            root.Left.Right.Right = new TreeNode(15);


            Console.WriteLine("**BINARY TREE**");
            //Inorder Traversals
            string treeInOrder = string.Join(",", root.InorderTraversal(root));
            Console.WriteLine($"InOrder Recursive: [{treeInOrder}]");
            //  (left --> root --> right)
            string treeInorderIter = string.Join(",", root.IterativeInorderTraversal(root));
            Console.WriteLine($"InOrder Iterative: [{treeInorderIter}]");

            Console.WriteLine("\r");

            //PreOrder Traversals
            string treePreOrder = string.Join(",", root.PreOrderTraversal(root));
            Console.WriteLine($"PreOrder Recursive: [{treePreOrder}]");
            // (root --> left --> right)
            string treePreOrderIter = string.Join(",", root.IterativePreOrderTraversal(root));
            Console.WriteLine($"PreOrder Iterative: [{treePreOrderIter}]");

            Console.WriteLine("\r");

            //PostOrder Traversals
            string treePostOrder = string.Join(",", root.PostOrderTraversal(root));
            Console.WriteLine($"PostOrder Recursive: [{treePostOrder}]");
            // (left --> right --> root)
            string treePostOrderIter = string.Join(",", root.IterativePostOrderTraversal(root));
            Console.WriteLine($"PostOrder Iterative: [{treePostOrderIter}]");

            Console.WriteLine("\n");

            // Comparisons
            // P - Tree
            TreeNode pRoot = new TreeNode(1);
            pRoot.Left = new TreeNode(2);
            pRoot.Right = new TreeNode(3);

            // Q - Tree
            TreeNode qRoot = new TreeNode(1);
            qRoot.Left = new TreeNode(2);
            qRoot.Right = new TreeNode(3);
            Console.WriteLine(pRoot.IsSameTree(pRoot, qRoot));


            // Is Symmetric?
            /*
                           1
                 		 /   \
                 	    2     2
                      /  \   /  \
                     3   4  4    3             
            */

            // Main Root
            TreeNode symmRoot = new TreeNode(1);

            // 1st Level
            symmRoot.Left = new TreeNode(2);
            symmRoot.Right = new TreeNode(2);

            // 2nd Level
            symmRoot.Left.Left = new TreeNode(3);
            symmRoot.Left.Right = new TreeNode(4);
            symmRoot.Right.Right = new TreeNode(3);
            symmRoot.Right.Left = new TreeNode(4);

            Console.WriteLine($"Symmetric: {symmRoot.IsSymmetric(symmRoot)}");
            Console.WriteLine($"IsBalanced: {symmRoot.IsBalanced(symmRoot)}");
            Console.WriteLine($"Minimum Depth: {symmRoot.MinDepth(symmRoot)}");


            // Path Sum
            /*
                        5
                       / \
                      4   8
                     /   / \
                    11  13  4
                   /  \      \
                  7    2      1            
            */

            // Main Root
            TreeNode rootSum = new TreeNode(5);

            // 1st Level
            rootSum.Left = new TreeNode(4);
            rootSum.Right = new TreeNode(8);

            // 2nd Level
            rootSum.Left.Left = new TreeNode(11);
            rootSum.Right.Left = new TreeNode(13);
            rootSum.Right.Right = new TreeNode(4);

            // 3rd Level
            rootSum.Left.Left.Left = new TreeNode(7);
            rootSum.Left.Left.Right = new TreeNode(2);
            rootSum.Right.Right.Right = new TreeNode(1);

            Console.WriteLine(rootSum.HasPathSum(rootSum, 22));

            string combinedLargest = string.Join(",",rootSum.LargestValues(rootSum));
            Console.WriteLine($"[{combinedLargest}]");

            string combinedPaths = string.Join(", ", rootSum.BinaryTreePaths(rootSum));
            Console.WriteLine($"[{combinedPaths}]");


            //  Complete Tree
            /*
                      1
                     / \
            	 	2 	3
            	   /\  /
             	  4  5 6
            */

            // Main Root
            TreeNode completeRoot = new TreeNode(1);

            // 1st Level
            completeRoot.Left = new TreeNode(2);
            completeRoot.Right = new TreeNode(3);

            // 2nd Level
            completeRoot.Left.Left = new TreeNode(4);
            completeRoot.Left.Right = new TreeNode(5);
            completeRoot.Right.Left = new TreeNode(6);

            // Count Nodes of a Complete Tree
            Console.WriteLine($"Nodes: {completeRoot.CountNodes(completeRoot)}");

            // Sum of all Left leaf nodes
            Console.WriteLine($"Left Leaves Sum: {completeRoot.SumOfLeftLeaves(completeRoot)}");

            // Level Order
            List<List<int>> levelOrderResult = completeRoot.LevelOrder(completeRoot);
            List<string> resultString = new List<string>();
            
            foreach (var level in levelOrderResult)
            {
                resultString.Add($"[{string.Join(", ", level)}]");
            }
            string combinedLevels = string.Join(", ", resultString);
            Console.WriteLine($"Level Order Traversal: [{combinedLevels}]");
        }
    }

    public class TreeNode
    {
        public int Data { get; set; }
        public TreeNode Left { get; set; }
        public TreeNode Right { get; set; }

        public TreeNode(int data)
        {
            Data = data;
            Left = null;
            Right = null;
        }

        //Recursive
        public int[] InorderTraversal(TreeNode root)
        {
            List<int> result = new List<int>();

            if (root == null)
            {
                return result.ToArray();
            }

            if (root.Left != null)
            {
                result.AddRange(InorderTraversal(root.Left));
            }

            result.Add(root.Data);

            if (root.Right != null)
            {
                result.AddRange(InorderTraversal(root.Right));
            }

            return result.ToArray();
        }

        //Iterative
        public int[] IterativeInorderTraversal(TreeNode root)
        {
            List<int> result = new List<int>();

            if (root == null)
            {
                return result.ToArray();
            }

            Stack<TreeNode> stack = new Stack<TreeNode>();
            TreeNode current = root;

            while (current != null || stack.Count > 0)
            {
                while (current != null)
                {
                    stack.Push(current);
                    current = current.Left;
                }

                current = stack.Pop();
                result.Add(current.Data);
                current = current.Right;
            }
            return result.ToArray();
        }

        //Recursive
        public int[] PreOrderTraversal(TreeNode root)
        {
            List<int> result = new List<int>();
            if (root == null)
            {
                return result.ToArray();
            }

            result.Add(root.Data);

            if (root.Left != null)
            {
                result.AddRange(PreOrderTraversal(root.Left));
            }

            if (root.Right != null)
            {
                result.AddRange(PreOrderTraversal(root.Right));
            }

            return result.ToArray();
        }

        //Iterative
        public int[] IterativePreOrderTraversal(TreeNode root)
        {
            List<int> result = new List<int>();

            if (root == null)
            {
                return result.ToArray();
            }

            Stack<TreeNode> stack = new Stack<TreeNode>();
            TreeNode current = root;

            while (current != null || stack.Count > 0)
            {
                while (current != null)
                {
                    result.Add(current.Data);
                    stack.Push(current);
                    current = current.Left;
                }

                current = stack.Pop();
                current = current.Right;
            }
            return result.ToArray();
        }

        //Recursive
        public int[] PostOrderTraversal(TreeNode root)
        {
            List<int> result = new List<int>();

            if (root == null)
            {
                return result.ToArray();
            }

            if (root.Left != null)
            {
                result.AddRange(PostOrderTraversal(root.Left));
            }

            if (root.Right != null)
            {
                result.AddRange(PostOrderTraversal(root.Right));
            }

            result.Add(root.Data);

            return result.ToArray();
        }

        //Iterative
        public int[] IterativePostOrderTraversal(TreeNode root)
        {
            List<int> result = new List<int>();

            if (root == null)
            {
                return result.ToArray();
            }

            Stack<TreeNode> stack = new Stack<TreeNode>();
            TreeNode current = root;
            TreeNode prev = null;

            while (current != null || stack.Count > 0)
            {
                while (current != null)
                {
                    stack.Push(current);
                    current = current.Left;
                }

                current = stack.Peek();

                if (current.Right == null || current.Right == prev)
                {
                    result.Add(current.Data);
                    stack.Pop();
                    prev = current;
                    current = null;
                }
                else
                {
                    current = current.Right;
                }
            }
            return result.ToArray();
        }

        public bool IsSameTree(TreeNode p, TreeNode q)
        {
            if (p == null && q == null)
            {
                return true;
            }

            if (p == null || q == null)
            {
                return false;
            }
            return p.Data == q.Data && IsSameTree(p.Left, q.Left) && IsSameTree(p.Right, q.Right);
        }

        public bool IsSymmetric(TreeNode root)
        {
            bool IsMirror(TreeNode left, TreeNode right)
            {
                if (left == null && right == null)
                {
                    return true;
                }
                if (left == null || right == null)
                {
                    return false;
                }
                return (left.Data == right.Data) && IsMirror(left.Left, right.Right) && IsMirror(left.Right, right.Left);
            }
            if (root == null)
            {
                return true;
            }
            return IsMirror(root.Left, root.Right);
        }

        public bool IsBalanced(TreeNode root)
        {
            Tuple<bool, int> CheckBalance(TreeNode node)
            {
                // An empty tree is balanced and has a height of 0.
                if (node == null)
                {
                    return new Tuple<bool, int>(true, 0);
                }

                // Check if the left subtree is balanced and get its height.
                bool isLeftBalanced = CheckBalance(node.Left).Item1;
                int leftHeight = CheckBalance(node.Left).Item2;

                if (isLeftBalanced == false)
                {
                    return new Tuple<bool, int>(false, 0);
                    // If the left subtree is not balanced, the whole tree is unbalanced.
                }

                bool isRightBalanced = CheckBalance(node.Right).Item1;
                int rightHeight = CheckBalance(node.Right).Item2;

                if (isRightBalanced == false)
                {
                    return new Tuple<bool, int>(false, 0);
                }

                bool isCurrentBalanced = Math.Abs(leftHeight - rightHeight) <= 1;
                int currentHeight = Math.Max(leftHeight, rightHeight) + 1;

                return new Tuple<bool, int> (isCurrentBalanced, currentHeight);
            }

            bool isBalanced = CheckBalance(root).Item1;
            return isBalanced;
        }

        // MinDepth till we reach a Leaf Node
        public int MinDepth (TreeNode root)
        {
            // Breadth-First Search
            if (root == null)
            {
                return 0;
            }
            Queue<Tuple<TreeNode, int>> queue = new Queue<Tuple<TreeNode, int>>();
            queue.Enqueue(new Tuple<TreeNode, int>(root, 1));

            while (queue.Count != 0)
            {
                var current = queue.Dequeue();
                TreeNode node = current.Item1;
                int depth = current.Item2;

                if (node.Left == null && node.Right == null)
                {
                    return depth;
                }

                if (node.Left != null)
                {
                    queue.Enqueue(new Tuple<TreeNode, int>(node.Left, depth + 1));
                }

                if (node.Right != null)
                {
                    queue.Enqueue(new Tuple<TreeNode, int>(node.Right, depth + 1));
                }
            }
            return 0;
        }

        public bool HasPathSum(TreeNode root, int targetSum)
        {
            if (root == null)
            {
                return false;
            }

            targetSum -= root.Data;

            if (root.Left == null && root.Right == null)
            {
                return targetSum == 0;
            }
            return HasPathSum(root.Left, targetSum) || HasPathSum(root.Right, targetSum);
        }

        public int[] LargestValues(TreeNode root)
        {
            List<int> result = new List<int>();

            if (root == null)
            {
                return result.ToArray();
            }

            Queue<TreeNode> queue = new Queue<TreeNode>();
            queue.Enqueue(root);

            while (queue.Count > 0)
            {
                float levelMax = float.NegativeInfinity;
                int levelSize = queue.Count;

                for (int i = 0; i < levelSize; i++)
                {
                    TreeNode node = queue.Dequeue();
                    levelMax = Math.Max(levelMax, node.Data);

                    if (node.Left != null)
                    {
                        queue.Enqueue(node.Left);
                    }

                    if (node.Right != null) 
                    {
                        queue.Enqueue(node.Right);
                    }
                }
                result.Add((int)levelMax);
            }
            return result.ToArray();
        }

        public string[] BinaryTreePaths(TreeNode root)
        {
            List<string> result = new List<string>();

            void dfs(TreeNode node, List<string> currentPath)
            {
                if (node == null)
                {
                    return;
                }

                currentPath.Add(node.Data.ToString());

                if (node.Left == null && node.Right == null)
                {
                    result.Add(string.Join("-> ", currentPath));
                }

                dfs(node.Left, currentPath);
                dfs(node.Right, currentPath);

                currentPath.RemoveAt(currentPath.Count - 1);
            }

            List<string> current = new List<string>();
            dfs(root, current);

            return result.ToArray();
        }

        public int CountNodes(TreeNode root)
        {
            if (root == null)
            {
                return 0;
            }

            int depthLeft = 0;
            int depthRight = 0;
            TreeNode left = root;
            TreeNode right = root;

            while (left != null)
            {
                left = left.Left;
                depthLeft++;
            }

            while (right != null)
            {
                right = right.Right;
                depthRight++;
            }

            if (depthLeft == depthRight)
            {
                return (int)(Math.Pow(2, depthLeft) - 1);
            }
            else
            {
                int leftCount;
                if (root.Left != null) {
                    leftCount = CountNodes(root.Left);
                } else {
                    leftCount = 0;
                }

                int rightCount;
                if (root.Right != null) {
                    rightCount = CountNodes(root.Right);
                } else {
                    rightCount = 0;
                }
                return 1 + leftCount + rightCount;
            }
        }

        public int SumOfLeftLeaves(TreeNode root)
        {
            if (root == null)
            {
                return 0;
            }

            int leftSum = 0;

            if (root.Left != null && root.Left.Left == null && root.Left.Right == null)
            {
                leftSum += root.Left.Data;
            }

            if (root.Left != null) {
                leftSum += SumOfLeftLeaves(root.Left);
            } else {
                leftSum += 0;
            }

            if (root.Right != null)
            {
                leftSum += SumOfLeftLeaves(root.Right);
            } else {
                leftSum += 0;
            }

            return leftSum;
        }

        public List<List<int>> LevelOrder(TreeNode root)
        {
            List<List<int>> result = new List<List<int>>();
            if (root == null)
            {
                return result;
            }

            Queue<TreeNode> queue = new Queue<TreeNode>();
            queue.Enqueue(root);

            while (queue.Count > 0)
            {
                List<int> level = new List<int>();
                int levelSize = queue.Count;

                for (int _ = 0; _ < levelSize; _++)
                {
                    TreeNode node = queue.Dequeue();
                    level.Add(node.Data);

                    if (node.Left != null)
                    {
                        queue.Enqueue(node.Left);
                    }

                    if (node.Right != null)
                    {
                        queue.Enqueue(node.Right);
                    }
                }
                result.Add(level);
            }
            return result;
        }


    }
}
